rtrlib/aspa/aspa_private.h

```diff
@@ -41,82 +114,122 @@ struct aspa_store_node {
  * @param[in,out] dst The destination table. Existing records associated with the socket are replaced.
  * @param[in,out] src The source table.
  * @param[in,out] rtr_socket The socket the records are associated with.
- * @param notify_dst A boolean value determining whether to notify the destination tables' clients.
- * @param notify_src A boolean value determining whether to notify the source tables' clients.
+ * @param notify_dst A boolean value determining whether to notify the destination table's clients.
+ * @param notify_src A boolean value determining whether to notify the source table's clients.
  */
 enum aspa_status aspa_table_src_replace(struct aspa_table *dst, struct aspa_table *src, struct rtr_socket *rtr_socket,
 					bool notify_dst, bool notify_src);
 
-// MARK: - Updating
-
-/**
- * @brief An enum describing the type of operation the ASPA table should perform using any given ASPA record.
- */
-enum aspa_operation_type {
-	/** The existing record, identified by its customer ASN, shall be withdrawn from the ASPA table. */
-	ASPA_REMOVE = 0,
-
-	/** The new record, identified by its customer ASN, shall be added to the ASPA table. */
-	ASPA_ADD = 1
-};
+// MARK: - Swap-In Update Mechanism
 
 /**
  * @brief A struct describing a specific type of operation that should be performed using the attached ASPA record.
  * @param index A value uniquely identifying this operation's position within the array of operations.
  * @param type The operation's type.
+ * @param skip A boolean value indicating whether this operation has been skipped while creating the update structure.
  * @param record The record that should be added or removed.
+ * @param is_no_op A boolean value determining whether this operation is part of a pair of 'add $CAS' and 'remove $CAS' operations that form a no-op.
  */
 struct aspa_update_operation {
 	size_t index;
 	enum aspa_operation_type type;
 	struct aspa_record record;
+	bool is_no_op;
 };
 
 /**
- * @brief ASPA update finalization arguments
- * @param unused_provider_arrays Array of pointers to now unused provider sets.
- * @param unused_provider_array_len Number of unused provider sets.
+ * @brief Computed ASPA update.
  */
-struct aspa_update_finalization_args {
-	uint32_t **unused_provider_arrays;
-	size_t unused_provider_array_len;
-#ifndef ASPA_UPDATE_IN_PLACE
-	struct aspa_array **old_array;
+struct aspa_update {
+	struct aspa_table *table;
+	struct aspa_update_operation *operations;
+	size_t operation_count;
```

Comment:
```suggestion
	size_t operation_count;
	struct aspa_update_operation *operations;
```
first count, then pointer to element. might cause better struct packing alignment.

Reply:
We're targeting 64/32b, so shouldn't they either both be 32b or 64b?